### PR TITLE
Define preview-aware and page-aware cache tags

### PR DESCRIPTION
**Pull request by RealOrangeOne**
Fixes https://github.com/wagtail/wagtail/issues/5074

This PR adds 2 new tags:

- `{% wagtailcache %}`: Functions identical to Django's `{% cache %}`, however does nothing during preview requests (always cache misses, never updates cache)
- `{% wagtailpagecache %}`: An extension of `{% wagtailcache %}`, but automatically varies its cache based on the current page and site. The tag intentionally makes assumptions to aid ease of use. If the user needs something more custom, it's basically the same as `{% wagtailcache page.id site.id ... %}`

_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [x] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    -   [ ] **Please list the exact browser and operating system versions you tested**:
    -   [ ] **Please list which assistive technologies [^3] you tested**:
-   [ ] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

Opening PR to gauge feedback. Will write relevant documentation soon.

### CODE REVIEW

**Comment by ababic** [comment]
@RealOrangeOne You may find `wagtail.coreutils.get_dummy_request()` useful here :)

File: docs/advanced_topics/performance.md
Comment (tomdyson) [inline docs/advanced_topics/performance.md:107]: "This makes it possible." ?

File: docs/advanced_topics/performance.md
Comment (tomdyson) [inline docs/advanced_topics/performance.md:125]: "the cache key for" - for what?

**Comment by ababic** [comment]
@RealOrangeOne was just looking at this again in the context of some menu generation issues we had recently on a client project, and wondered if `wagtailpagecache` over-complicating things by including the current site ID, since each page can only canonically belong to a single site, which is unlikely to change.

If the intention is to cater for cases where a page moves from one site to another, perhaps using the page's `url_path` value might be simpler? That would have the added bonus of invalidating when url changes (for the current and descendant pages) are likely:
- The page's slug (or that of an ancestor) changed
- The page was moved to a different part of the tree on the same site
- The page was moved to a different site (unless the root pages of the sites were swapped around, which is highly unlikely)

Another potential gain here might be to prefix the cache keys in some way to make unintentional clashes less likely? For example: `"Page.id:3"` and `"Site.id:1"` (or `"Page.url_path:/home/my-parent/my-slug/"`, if the above change were made). Admittedly, this is less of a concern, but is easy to do, and can't imagine it would hurt?

File: wagtail/models/__init__.py
Comment (ababic) [reply wagtail/models/__init__.py:2325]: You know what would be cool? If the components were gathered by another method (e.g. `get_cache_key_components()`) That would overrides easy without having to repeat all of the hashing logic etc.

Comment (RealOrangeOne) [reply wagtail/models/__init__.py:2325]: > If the components were gathered by another method

oooohhh, love this idea!

File: wagtail/coreutils.py
Comment (jacobtoppm) [inline wagtail/coreutils.py:545]: Given we're letting people customise the cache key components, do we need to account for that here/when mentioning this fn in docs?

Comment (jacobtoppm) [reply wagtail/coreutils.py:545]: Ah, ignore me, just noticed the `cache_key` is already in here

File: wagtail/models/__init__.py
Comment (jacobtoppm) [inline wagtail/models/__init__.py:2318]: Just want to flag that it is possible to programatically update a live page without increasing this (and it could trip up people trying to make urgent updates) - is it worth noting that in the docs, or adding an explicit cache purge somewhere?

Comment (RealOrangeOne) [reply wagtail/models/__init__.py:2318]: Explicit cache purging is difficult, as it'd get difficult to know whether you're clearing the cache before or after the update, or even what the final cache key ended up being. I agree that we should document it.

**Comment by ababic** [comment]
@RealOrangeOne I was thinking about this from the context of adoption, and making it feel more like a natural replacement for the Django version, and had a couple of thoughts:

In Django, you add `{% load cache %}` to templates to access the tags. Maybe we should have a separate module for these too, e.g. `{% load wagtail_cache %}`. This may help with ongoing maintenence too, I think.

If we load these tags explicitly, maybe we can lose the `wagtail` prefix from our version of the tag names? This way, devs might just need to do a mass find/replace for `{% load cache %}` -> `{% load wagtail_cache %}` to make use of this.

To top it all off, maybe we can change the page-specific version to something slightly more obvious, e.g. `{% cache_for_page %}`. 

Naming things ;-P

**Comment by ababic** [comment]
Coolio. Happy to take your lead on the tag naming 😃

**Comment by ababic** [comment]
The only outstanding thing from my perspective is moving the cache-related stuff out of `wagtailcore_tags.py` and into a new module.

File: CHANGELOG.txt
Comment (RealOrangeOne) [inline CHANGELOG.txt:8]: I've marked this as 5.1, as presumably we're too late for 5.0?

File: docs/advanced_topics/performance.md
Comment (RealOrangeOne) [inline docs/advanced_topics/performance.md:97]: I've had to name the library `wagtail_cache` to avoid conflicting with [`wagtail-cache`](https://github.com/coderedcorp/wagtail-cache)'s `wagtailcache_tags`.

File: wagtail/templatetags/wagtail_cache.py
Comment (Tijani-Dia) [inline wagtail/templatetags/wagtail_cache.py:63]: Nitpick: I think we can omit the `return` statement here.

**Comment by ababic** [comment]
Thanks for the updates @RealOrangeOne. I may have some time on this Monday or Tuesday to give this a test run, but I do have a new project kicking off this week, which I'll be getting my head around.

If this needs to be throughly tested by a core team member before merging anyhow, I wonder how much value another review from me will bring 🤷

File: docs/advanced_topics/performance.md
Comment (ababic) [inline docs/advanced_topics/performance.md:101]:
```suggestion
The `{% wagtailcache %}` tag functions similarly to Django's `{% cache %}` tag, but will neither cache or serve cached content when previewing a page (or other model) in Wagtail.
```

File: docs/advanced_topics/performance.md
Comment (ababic) [inline docs/advanced_topics/performance.md:146]:
```suggestion
It's often necessary to cache a value based on an entire page, rather than a specific value. For this, {attr}`~wagtail.models.Page.cache_key` can be used to get a unique value for the state of a page. Should something about the page change, so will its cache key. You can also use the value to create longer, more specific keys when using Django's caching framework directly. For example:
```

File: docs/advanced_topics/performance.md
Comment (ababic) [inline docs/advanced_topics/performance.md:158]:
```suggestion
To modify the cache key, such as including a custom model field value, you can override {attr}`~wagtail.models.Page.get_cache_key_components`:
```

File: docs/advanced_topics/performance.md
Comment (ababic) [inline docs/advanced_topics/performance.md:167]:
```suggestion
Manually updating a page might not result in a change to its cache key, unless the default component field values are modified directly. To be sure of a change in the cache key value, try saving the changes to a `Revision` instead, and then publishing it.
```

File: wagtail/tests/tests.py
Comment (ababic) [inline wagtail/tests/tests.py:610]: Would it be worth adding a `test_invalid_use()` case here, just to cover off the lines that Codecov is complaining about?

**Review by ababic — COMMENTED** [review]
I have a added a few suggestions here for you consideration, @RealOrangeOne.

File: wagtail/tests/tests.py
Comment (RealOrangeOne) [reply wagtail/tests/tests.py:610]: We could, but those lines are specifically copied from Django. If the upstream lines change, so should these, regardless of what a test says.

File: docs/advanced_topics/performance.md
Comment (ababic) [reply docs/advanced_topics/performance.md:101]: A couple of very small things, really;
— 'identically' didn't feel quite right to me when immediately followed by a description of a thing that is different
— The 'does nothing' description didn't feel very useful, but the stuff in brackets did, so thought we might as well just explain the difference in a single sentence for better 'flow' .

File: wagtail/tests/tests.py
Comment (ababic) [reply wagtail/tests/tests.py:610]: @RealOrangeOne just to clarify, I mean to get coverage on the token/argument parsing lines.

File: docs/advanced_topics/performance.md
Comment (RealOrangeOne) [inline docs/advanced_topics/performance.md:97]: Note: I tried to use intersphinx to reference this Django docs page, but couldn't quite get it working. If someone else fancies giving it a go, please do!

**Review by jacobtoppm — APPROVED** [review]
LGTM, thanks for all your work on this!

**Comment by lb-** [comment]
Squashed, rebased, added release notes and will merge in. I will merge this into 5.2 though as we did not merge before the RC, unless there is push back on that I can change it to a 5.1 item.

**Pull request merged**
